pkg/cloudprovider/aws.go

```diff
@@ -1,37 +1,338 @@
 package cloudprovider
 
 import (
+	"errors"
+	"fmt"
 	"net"
 
+	awsapi "github.com/aws/aws-sdk-go/aws"
+	awscredentials "github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
 	corev1 "k8s.io/api/core/v1"
+	utilnet "k8s.io/utils/net"
 )
 
 const (
 	aws = "aws"
 )
 
-// AWS implements the API wrapper for talking
-// to the AWS cloud API
+var NoNetworkInterfaceError error = errors.New("no retrievable network interface")
+
+var (
+	awsIPv4FilterKey = "network-interface.addresses.private-ip-address"
+	awsIPv6FilterKey = "network-interface.ipv6-addresses.ipv6-address"
+)
+
+// AWS implements the API wrapper for talking to the AWS cloud API
 type AWS struct {
-	cloud CloudProvider
+	CloudProvider
+	region string
+	client *ec2.EC2
+}
+
+// AWSWaitInput is the required input for the AWS EC2 Wait API call (WaitUntilInstanceRunning).
+// Unfortunately the API only handles equality assertion: so on delete we can't
+// specify and assert that the IP which is being removed is completely removed,
+// we are forced to do the inverse, i.e: assert that all IPs except the IP being
+// removed are there (which should anyways be the case). Hence we might risk
+// having a very small time window where the state is incorrect. This can be
+// problematic for update operation. The assumption here is that AWS EC2 will be
```

Comment:
I don't understand how this issue leads to "a very small time window where the state is incorrect" or why that's only a problem for updates.